### ISSUE TITLE
How to deal with functools.singledispatch?

### ISSUE THREAD

**Issue by madig** [microsoft/pyright#988]
**Describe the bug**
functools.singledispatch relies on type signatures differently than what pylance expects.

**To Reproduce**
Adapted from https://docs.python.org/3/library/functools.html#functools.singledispatch:

```python
from functools import singledispatch
from typing import Any


@singledispatch
def fun(arg: Any, verbose: bool = False):
    if verbose:
        print("Let me just say,", end=" ")
    print(arg)


@fun.register
def _(arg: list, verbose: bool = False):
    if verbose:
        print("Enumerate this:")
    for i, elem in enumerate(arg):
        print(i, elem)
```

Pylance complains about the second function's signature:
```
Type of parameter "arg" is partially unknown
  Parameter type is "list[Unknown]"
```

Making it e.g. a `List[int]` makes it type check but breaks singledispatch, which does just a simple `isinstance` check. So, uh, what do?

**Expected behavior**
This should type check without errors I guess? Or not?

**VS Code extension**
VSCode 1.48.2
Pylance 2020.8.3
Python 2020.8.105369

**Comment by erictraut** [comment]
I presume that you have "strict" type checking enabled. The "reportUnknownParameterType" diagnostic rule is disabled by default if the type checking mode is set to "basic".

I'm not familiar with the details of functools.singledispatch. It appears that it's using the type annotation of the parameter at runtime, which is pretty strange to begin with. The fact that it's complaining about "List[int]" not being a class seems to be a bug in the implementation of the decorator.

A few ideas about how to work around this:
1. Fall back on old comment-based parameter type annotations
```
@fun.register
def _(
    arg,  # type: List[int]
    verbose: bool = False,
):
    if verbose:
        print("Enumerate this:")
    for i, elem in enumerate(arg):
        print(i, elem)
```

2. Use a `# type: ignore` comment to suppress Pyright's error message:
```
@fun.register
def _(arg: list, verbose: bool = False):  # type: ignore
    typed_arg: List[int] = arg
    if verbose:
        print("Enumerate this:")
    for i, elem in enumerate(typed_arg):
        print(i, elem)
```

3. Disable the specific diagnostic rules for that file by placing the following comment at the top:
```
# The following diagnostic rules need to be disabled because the @fun.register
# decorator generates runtime exceptions when a proper type annotation is used
# for parameters.
# pyright: reportUnknownParameterType=false, reportMissingTypeArgument=false
```

**Comment by layday** [comment]
The decorator is correct, `List` is not a class at runtime - you can verify this by running `isinstance(typing.List, type)`.  Personally, I would pass the type to `register` which takes precedence over the annotation:

```py
from functools import singledispatch
from typing import Any, List


@singledispatch
def fun(arg: Any, verbose: bool = False):
    if verbose:
        print("Let me just say,", end=" ")
    print(arg)


@fun.register(list)
def _(arg: List[str], verbose: bool = False):
    if verbose:
        print("Enumerate this:")
    for i, elem in enumerate(arg):
        print(i, elem)
```

But this is all quite pointless from a type checking perspective since Pyright won't know when to use `Any` or `List[str]` - it only really works if all the functions return the same type object and if the parent function dispatches on `Any`.  In addition the `singledispatch` decorator obscures the parameters of the parent function.

**Comment by madig** [comment]
Thanks for the input. So, it's messy. I'll be back later with a question about how to make strict Pylance not recognize all variants of the singledispatch as `NoReturn` with all the non-reachableness that entails if the default functions is marked to return `NoReturn`.

**Comment by madig** [comment]
Ok, here goes again:

```python
from functools import singledispatch
from typing import Any, List, NoReturn


@singledispatch
def fun(arg: Any) -> NoReturn:
    raise ValueError("no.")


@fun.register(list)
def _(arg: List[str]) -> str:
    return str(arg)
```

```
Argument of type "(arg: List[str]) -> str" cannot be assigned to parameter of type "(*args: Unknown, **kwargs: Unknown) -> NoReturn"
  Type "(arg: List[str]) -> str" cannot be assigned to type "(*args: Unknown, **kwargs: Unknown) -> NoReturn"
    Function return type "str" is not compatible with type "NoReturn"
      "str" is incompatible with "NoReturn"
```

We use this construct here: https://github.com/fonttools/fonttools/blob/4.14.0/Lib/fontTools/misc/plistlib.py#L374-L388. The default function should raise, only registered handlers actually handle data.

**Comment by teichert** [comment]
@madig: thanks for getting me playing with the order the `@overload` vs `@singledispatch`.  Didn't you get errors at runtime because the `@overload`-decorated version came last, wiping out your actual implementation?

A workaround is to use singledispatch or [another library](https://github.com/coady/multimethod) to accomplish the actual overloading *using a different name* and then use `@overload` from typing to annotate the possible method signatures (mypy would know them from singledispatch, but apparently vscode doesn't [I'm actually using pylance rather than pyright]), and then implement the function with a single call to the overloaded one.

This is the only way that I have found to get vscode intellisense to work while pleasing mypy and without breaking runtime:

```python
from functools import singledispatch
from typing import List, overload


@singledispatch
def _fun(arg) -> str:
    return "no"


@_fun.register(list)
def _(vals: List[str]) -> str:
    return str(vals)


@overload
def fun(arg) -> str: ...


@overload
def fun(vals: List[str]) -> str: ...


def fun(*args, **kwargs):
    return _fun(*args, **kwargs)


def test_overload():
    assert fun(78) == 'no'
    assert fun(['test', 'overload']) == "['test', 'overload']"

```

**Comment by ShivnarenSrinivasan** [comment]
@teichert thanks a bunch for this—really helpful.

On a side note, generics like `List` from the `typing` module are deprecated—you can just use the primitive `list`.
